apps/massa_proxy/lib/massa_proxy/runtime/grpc/protocol/action/stream_handler.ex

```diff
@@ -2,42 +2,120 @@ defmodule MassaProxy.Runtime.Grpc.Protocol.Action.Stream.Handler do
   """
-  require Logger
-  alias GRPC.Server
+
   alias Google.Protobuf.Any
+  alias Cloudstate.{Action.ActionCommand, Metadata}
+  alias Cloudstate.Action.ActionProtocol.Stub, as: ActionClient
+
+  import MassaProxy.Util, only: [get_connection: 0, get_type_url: 1]
+
+  def handle_streamed(%{stream: stream, message: message} = payload) do
+    messages = build_stream(message, payload)
+
+    with {:ok, conn} <- get_connection(),
+         client_stream = ActionClient.handle_streamed(conn),
+         :ok <- client_stream |> run_stream(messages) |> Stream.run(),
+         {:ok, consumer_stream} <- GRPC.Stub.recv(client_stream) do
+      consumer_stream
+      |> Stream.each(fn {:ok, r} -> GRPC.Server.send_reply(stream, unpack_response(r)) end)
+      |> Stream.run()
+    else
+      {:error, _reason} = err -> err
+    end
+  end
+
+  def handle_stream_in(%{message: message} = payload) do
+    messages = build_stream(message, payload)
+
+    with {:ok, conn} <- get_connection(),
+         client_stream = ActionClient.handle_streamed_in(conn),
+         task_result <- run_stream(client_stream, messages),
+         :ok <- accumlate_stream_result(task_result),
+         {:ok, response} <- GRPC.Stub.recv(client_stream) do
+      unpack_response(response)
+    else
+      {:error, _reason} = err -> err
+    end
+  end
+
+  def handle_stream_out(%{stream: stream} = payload) do
+    message =
+      ActionCommand.new(
+        service_name: payload.service_name,
+        name: payload.original_method,
+        payload:
+          Any.new(
+            type_url: get_type_url(payload.input_type),
+            value: payload.input_type.encode(payload.message)
+          ),
+        metadata: Metadata.new()
+      )
+
+    with {:ok, conn} <- get_connection(),
+         {:ok, client_stream} <- ActionClient.handle_streamed_out(conn, message, []) do
+      Stream.each(client_stream, fn {:ok, response} ->
+        GRPC.Server.send_reply(stream, unpack_response(response))
+      end)
+      |> Stream.run()
+    end
+  end
+
+  defp run_stream(client_stream, messages) do
+    Stream.map(messages, &send_stream_msg(client_stream, &1))
+  end
 
-  def handle_streamed(
-        %{
-          stream: stream,
-          input_type: input_type,
-          output_type: output_type
-        } = payload
-      ) do
-    Stream.each(stream, fn msg ->
-      Logger.info("Decode request from #{inspect(msg)}")
-      handle_streamed_message(stream, msg)
+  defp unpack_response(%{response: {_, response}}), do: response
+
+  defp accumlate_stream_result(results) do
+    Enum.reduce_while(results, :ok, fn
+      {:error, _reason} = err, _acc -> {:halt, err}
+      _, acc -> {:cont, acc}
     end)
-    |> Stream.run()
   end
 
-  def handle_stream_in(payload) do
+  defp build_stream(msgs, state) do
+    # Need to signal end_of_stream so attach a token to
+    # end of the stream so that we can call `GRPC.Stub.end_stream`
+    msgs
+    |> Stream.concat([:halt])
```

Comment:
What happens if the input stream is an infinite stream? gRPC can be used with infinite flows, that is, once a connection is established I can send an arbitrary number of messages without a predetermined end.

Reply:
It will enumerate the entire incoming stream before reaching `:halt`. Using the token is to signal `end_of_stream` in the case that we actually consumed the entire stream and we're done, otherwise the user function will wait for more input and eventually timeout.

If we're processing indefinitely we'll never reach  (or it will take a long time 😄 ) the `:halt` token but as we'll be sending the entire time the stream is *active*.

Reply:
I'm still in doubt whether this works with long streams, we can't assume that users will terminate the data stream for any reason. We will have to test this approach